PR#7075: fix repeated docstrings

Octachron opened the pull request:
This pull request hopefully fixes the problem of the repetition of documentation comments in ocamldoc-generated documentations. At least, it fixes all repetition and empty section title problems reported in [mantis:7075](http://caml.inria.fr/mantis/view.php?id=7075).

Octachron committed:
Test suite: check ocamldoc generated tex files

Basic safety net for ocaml manual

Octachron committed:
PR#7075: Update changes

lpw25 commented:
I think this patch probably correctly fixes the handling of documentation attributes in ocamldoc. However, I think that a better approach would be to avoid producing such attributes in ocamldoc. This could be done quite easily by having a ref in the lexer `handle_doc_comments` which can be set to `false` and means that the lexer never produces `DOCSTRING` tokens (producing `COMMENT` instead). This would have the added benefit of preventing the doc comment placement warnings (which are not 100% backwards compatible with ocamldoc) from being confusingly emitted when using ocamldoc.

gasche commented:
@Octachron thanks for working on this!

@lpw25 I have mixed feelings about disabling documentation comments parsing from ocamldoc (it feels a bit like admitting defeat, although in your case I'd rather guess: lack of time, currently, to prepare a real victory on this front), but it may be a good suggestion. I'm thinking of merging the present patch anyway to fix the reported issues (and get the testsuite coverage goodness), because I'm not sure anyone will be motivated to work on this further in the current release timeframe -- and I would rather get the correct behavior as seen as possible. Does that sound reasonable?

Octachron committed:
PR#7075: fix repeated documentation comments

This commit adds a hook in "parsing/lexer.ml{i,l}" to deactivate the generation of
docstring items. This hook is used by ocamldoc to avoid interferences between these
items and the ocamldoc documentation comments parser.

Octachron commented:
I did not think about deactivating the docstring items generation from the lexer. It also works quite well, see [this alternative branch](https://github.com/ocaml/ocaml/compare/trunk...Octachron:fix_repeated_docstrings_bis?expand=1). (Personally, I have no idea which approach is the best.)

gasche commented:
Leo knows best about this part obviously, so I'll defer to his decision.

In terms of design I'm not personally convinced by the "disabling documentation comments" approach (in fact I'd probably be in favor of pushing harder by removing the ocamldoc-specific code to parse comments and relying on the language-parser-support only, but probably not in the coming release), but in practice I'm fine with merging either patchsets for 4.03.

lpw25 commented:
> In terms of design I'm not personally convinced by the "disabling documentation comments" approach (in fact I'd probably be in favor of pushing harder by removing the ocamldoc-specific code to parse comments and relying on the language-parser-support only, but probably not in the coming release)

The code which currently doesn't work with documentation comments will just be deleted when we switch ocamldoc to using just the builtin documentation comment support -- once documentation comments are represented as attributes and can pass safely through the type-checker it just isn't needed any more -- so there is no real point in making it understand them. I prefer the lexer based solution because it is easier to see that it is correct and, as I said before, because it also avoids giving confusing warnings.

@Octachron If you push the alternative branch to this pull request I'll happily merge it.

Octachron commented:
Ok, I have replaced the original branch with the lexer alternative one.

alainfrisch commented on parsing/lexer.mll:
```diff
@@ -377,7 +378,11 @@ rule token = parse
+        if !handle_docstrings then
+          DOCSTRING (Docstrings.docstring s loc)
+        else
+          COMMENT (s, loc)
```
This might be irrelevant, but shouldn't the s be replaced by ("*" ^ s)?  Or, if ocamldoc simply ignores the comment content, simply drop the token (in the loop inner function under token)?

Octachron replied:
For the sake of coherence is indeed probably better to keep the leading "*". 
It does not affect ocamldoc, since ocamldoc reimplements all the comment parsing logic on its side. Droping the comment feels like a quite radical solution compared to just deactivating the special handling of docstring.

gasche replied:
I think @alainfrisch is worried about [MPR#6972](http://caml.inria.fr/mantis/view.php?id=6972) -- whatever you do, you should make sure that it is not an issue. (I guess neither Alain nor I looked at the ocamldoc issue in details enough to know whether that applies here.)

Octachron replied:
Right. Interaction with external tools (that would purposefully deactivate the special handling of docstring in the lexer) is certainly a good reason to add back the leading '*' start. (It might also be useful to check in the test suite that no comments are lost in cmt files.)

lpw25 commented:
I've looked at the patch and I think it's ready to merge. It seems there is now a conflict (maybe the changes file?). It'd be great if you could fix the conflict, if not I'll get around to doing it at some point.

Octachron commented:
The conflict was indeed some bug fixes moving around in the changes file. I have rebased the pull request to solve this conflict and squashed the two bug fix commits along the way.

lpw25 commented:
Thanks
